Tunable use of letter spacing for justification

poire-z opened the pull request:
See commit messages for details.

Discussion and screenshots at https://github.com/koreader/koreader/pull/5628#issuecomment-622354587

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/crengine/341)
<!-- Reviewable:end -->

poire-z commented:
(Some strange issue I just met, that I'll look at tomorrow: on some line in a book, 5% differs from 0%, but 15% shows just like 0% - while it should at least show like 5%...)

poire-z commented:
OK, I have a small edge case issue - which is explainable and we might live with it, but just detailing it for opinion.

(In all the examples below, we do only one loop, trying 1px as letter_spacing - there's no occasion where we need to use 2px. The `5` in the right margin means 5% and that we are using 1px letter_spacing.)

On a line, we may have words in multiple font sizes, eg., the `22` and the `x` in:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80861812-1a615180-8c71-11ea-9ec7-49f45b7aa80a.png)</kbd>

Depending on the % we set, we compute the max usable letter spacing as % of font size.
With 15%, the `22` can get 1px (shown above).
While with 5%, it can get 0px (while the other words in a larger font size can get 1px):
<kbd>![image](https://user-images.githubusercontent.com/24273478/80861832-54325800-8c71-11ea-8434-e9a98206bc08.png)</kbd>

No issue with that line, it works as expected (with 5%, adding 1px to the `22` would make it look more spreaded (compared to the other words where 1px does not look that spreaded), so it's fine we don't add 1px just to this `22`).

Now, with that other line, shown here with 0%:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80861931-d15dcd00-8c71-11ea-889d-ae41341fe3b5.png)</kbd>

With 5% (all words getting 1px, except that `22` getting 0px):
<kbd>![image](https://user-images.githubusercontent.com/24273478/80861959-ed616e80-8c71-11ea-8993-d6ba549ea26a.png)</kbd>

With 15%, first loop, all words getting 1px, and `22` too, we're sadly back to look like as with 0%:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80861990-0e29c400-8c72-11ea-9d25-de8908c9e1d7.png)</kbd>
because the `22` getting 1px makes us reach the limit where there's too much spreading and not enough space for spaces anymore, so we go back at using the previous letter spacing (at it's the first loop with letter_spacing=1px, we go at using 0px).

So, that's a limit/edgecase/badness with my algorithm :) But I can't think of any other way to do that.
(I browsed quite a few pages looking for other such cases, but couldn't find any, so it might very well be a rare situation.)
Any thoughts (some tweak in the algo I can't think of, some other algo, or "let's just live with that":) ?

Frenzie commented:
I'm not a lover of big letter spacing so it's a bit hard to opine on.

poire-z commented:
Yep, it's possible I won't use it either. But _medium_ letter spacing is quite ok :)

I guess I shouldnt use that tunable % as a variable in the loop to make out values with, but just as something for checking limits - so the loops make out the same values whatever that % is, they would just stop at a different iteration.

poire-z commented:
Anyway, as indeed it might not appeal to some people, and I think it should not mess with the Word Spacing toggle, it might have to be a new toggle (I won't say it _deserve_ one, but well, it has to be), any opinion on the 2nd part of that (https://github.com/koreader/koreader/pull/5628#issuecomment-622354587), small reorganising of the bottom menu:

> Anyway, I don't know if it should be another toggle, or integrated in the Word Spacing one. I initially thought combining it into Word Spacing, but it might hardcode our choice for this new setting, and people might not like that added spacing...
So, `Word Stretching off | small | large` ?
But that will make this bottom config menu have 6 items, which might cover a bunch of text and keep less of it to see the effects of all the settings... Any idea for reorganising that bottom menu (the Font size menu has only 2... but not obvious what we could move here - and it would make it different from the PDF one).

Frenzie commented:
Word & letter spacing can probably go in the font size thingy?

poire-z commented:
That's an idea.
I initially thought regular|bord and contrast could go into the font size panel.
And we would have hinting + kerning + word spacing + letter spacing(word stretching) in the next panel - which is where I would go tweaking may be more often (after I have choose a font size (and rarely touch boldness and contrast)). Having them in 2 panels would make me swap between them.
But then, the icon for that panel is the big bold **B**, and its content has nothing to do anymore with boldness :)

Although I think I hardly play with hinting and kerning :) So, may be the first panel (size + word spacing + letter spacing) would be the one most often used - and the next panel (boldness, contrast, hinting and kerning) would be the one less often used.
@NiLuJe : any thought?

Frenzie commented:
Maybe something like the Word line spacing thing?

![WINWORD_mDtJIZyhiq](https://user-images.githubusercontent.com/202757/80866875-6de39780-8c91-11ea-9fc9-1b4b30a7f856.png)

poire-z commented:
OK, I hopefully fixed my algorithm. (But I'm not convinced I'm done with it :)

> Maybe something like the Word line spacing thing?

To replace the **B** icon ? Dunno, and we don't have that icon yet, and we would lose the similarity with the PDF bottom panel.
Having spacing in the **Aa** panel would be fine, could relate to spacing between **A** and **a** :)
Boldness, contrast, hinting and kerning could relate to **B**, individual letters weight and quality (although kerning impact letters relations too).

poire-z commented:
@NiLuJe : do you think _hinting_ may change _advances_?
I'm wondering if toggling hinting really needs a re-rendering. If it may change advances and widths, it needs it (as some words could be pushed onto next line, changing a paragraph height).
If we're sure it doesn't change advances, we could avoid that.

NiLuJe commented:
> @NiLuJe : do you think _hinting_ may change _advances_?

IIRC, it definitely does.

NiLuJe commented:
And, yeah I definitely agree with splitting this to its own thing in the UI ;).

Moving stuff to the font size tab seems harmless enough to me ;).

poire-z commented:
OK, I guess hinting can change advances.

http://madig.github.io/freetype-web/documentation/glyphconventions/

> The glyph metrics are also influenced by the grid-fitting process:
> - The image’s width and height are altered. Even if this is only by one pixel, it can make a big difference at small pixel sizes.
> - The image’s bounding box is modified, thus modifying the bearings.
> - The advances must be updated. For example, the advance width must be incremented if the hinted bitmap is larger than the scaled one, to reflect the augmented glyph width.

poire-z commented:
The 2nd commit was all wrong, comparing glyph nums and cluster nums - cluster nums relate only to text indices, and glyph nums varies depending on the glyphs needed.
Witnessed while just adding the first version to xtext (which, for some reason, didn't exhibit any problem before and wouldn't need that change), where everything became a mess :/

poire-z commented:
Bold panel reduced from 5 (or 6 with the new one) to 4:

<kbd>![image](https://user-images.githubusercontent.com/24273478/80893050-44e1f780-8ccf-11ea-8071-1a1af6b6a9d0.png)</kbd>

Not thrilled by any of the following for the font panel.
Any preference or other ideas?
(Word spacing should eventually get its own `...` for fine tuning, so it may bring a bit more balance.)

<kbd>![image](https://user-images.githubusercontent.com/24273478/80893046-3d225300-8ccf-11ea-851c-a036fe9ef6e4.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/80893069-5aefb800-8ccf-11ea-995e-fd12cba595c5.png)</kbd>

<kbd>![image](https://user-images.githubusercontent.com/24273478/80893085-7ce93a80-8ccf-11ea-86dd-d0a26890bfea.png)</kbd>

NiLuJe commented:
@poire-z: Strong preference for the first option (i.e., size widget on top).

poire-z commented:
It's really not obvious to get a grasp of the relation between spacing vs stretching. It's fine when you play only with one - but playing with both, it's easy to wonder wtf's happening :)
Any idea for other adjectives than _small medium large_ for one of them (the fact that each grom from small to large in opposite direction might contribute to the confusion).

NiLuJe commented:
I'd honestly go with something like "Glyph Spacing" (or whatever the actual exact term is, because there's probably one ^^).

Frenzie commented:
@NiLuJe I think you just mean letter spacing, but since we're talking about dynamic letter spacing on top of the usual here that's probably not quite the right term.

poire-z commented:
`Space Reducing` or something about the result - rather than the method?

Frenzie commented:
For LaTeX, see https://ctan.org/pkg/microtype

poire-z commented:
Tsss, "Subliminal refinements towards typographical perfection" would have make a fancier title for this PR :)

Added a few micro-typography micro-fixes of micro-bugs:
- Fix first word after initial quotation mark/dash not getting the letter spacing - while still ensuring that first space from being fixed:
<kbd>![image](https://user-images.githubusercontent.com/24273478/80915271-7b605680-8d51-11ea-9753-da61c0c8a64b.png)</kbd>
- And fixed text selection/highlights to not include the added letter spacing at boundaries, and prevent multilines selection to have ragged left and right sides.

NiLuJe commented:
@poire-z:

> OK, I guess hinting can change advances.

Actually, I was just reminded that with the switch to `FT_RENDER_LIGHT`, there's no longer *any* hinting on the x axis, so, for all intents on purposes, horizontal metrics should be constant no matter the hinting mode ;).

poire-z commented:
Right, you mentionned that at https://github.com/koreader/koreader/pull/5553#issuecomment-548574784 :)

I'll see at making hinting changes no more needing a re-rendering (that's a lot of -ing in a sentence).

virxkane commented on crengine/src/lvdocview.cpp:
```diff
@@ -6135,11 +6137,25 @@ void LVDocView::propsUpdateDefaults(CRPropRef props) {
     props->setInt(PROP_FORMAT_SPACE_WIDTH_SCALE_PERCENT, p);
 
     p = props->getIntDef(PROP_FORMAT_MIN_SPACE_CONDENSING_PERCENT, DEF_MIN_SPACE_CONDENSING_PERCENT);
+    if (p<0)
+        p = 0;
+    if (p>20)
+        p = 20;
+    props->setInt(PROP_FORMAT_MIN_SPACE_CONDENSING_PERCENT, p);
```
So, PROP_FORMAT_MIN_SPACE_CONDENSING_PERCENT here can have values from 0 to 20, but default value 50.
Also in lvtextfm.cpp:
```cpp
void LFormattedText::setMinSpaceCondensingPercent(int minSpaceCondensingPercent)
{
    if (minSpaceCondensingPercent>=25 && minSpaceCondensingPercent<=100)
        m_pbuffer->min_space_condensing_percent = minSpaceCondensingPercent;
}
```
Copy-past issue?

Also in void LVDocView::createEmptyDocument():
```m_doc->setMinSpaceCondensingPercent(m_props->getIntDef(PROP_FORMAT_MIN_SPACE_CONDENSING_PERCENT, 50));``` used integer constant 50, but you already have macro DEF_MIN_SPACE_CONDENSING_PERCENT and so on.

poire-z replied:
> Copy-past issue?

Yep :/
We don't really use that lvdocview.cpp part in KOReader, so indeed, I must have updated it without paying much attention.
The lvtextfm.cpp is the part that have it right.

> used integer constant 50, but you already have macro DEF_MIN_SPACE_CONDENSING_PERCENT and so on

Indeed, but this line is from 2011 by buggins - the 3 others are mine and I just adapted to my surroundings :)

Fell free to fix all that, and I'll pick it.

poire-z replied:
OK, fixed that as part of #361 (3rd commit) which includes a 4th commit related to lvfntman.cpp that will need some adaptation on your side, but it shouldn't be much.

poire-z commented:
(Looks like Kobo's KEPUB renderer, when the inter-word space is becoming too large, adds some (possibly too large too) inter-letter spacing to the first word only:
![image](https://user-images.githubusercontent.com/24273478/88555307-ea370e00-d027-11ea-91c3-c2fd1c770d79.png)
https://www.mobileread.com/forums/showthread.php?t=331852
Which is kinda ugly - obviously caused there by hyphenation being disabled.)
